backend/controllers/prisonerProfile/prisonerCellHistory.js

```diff
@@ -42,27 +44,43 @@ module.exports = ({ oauthApi, prisonApi, logError, page = 0 }) => async (req, re
 
     const today = moment().format('YYYY-MM-DDTHH:mm:ss')
 
-    const cellData = cells
-      .sort((left, right) => sortByDateTime(right.assignmentDateTime, left.assignmentDateTime))
-      .filter(cell => cell.assignmentEndDateTime)
-      .map(cell => {
-        const agency = agencyData.find(agencyDetails => cell.agencyId === agencyDetails.agencyId)
-        return {
-          establishment: agency.description,
-          location: extractLocation(cell.description, cell.agencyId),
-          movedIn: cell.assignmentDateTime && formatTimestampToDateTime(cell.assignmentDateTime),
-          movedOut: cell.assignmentEndDateTime && formatTimestampToDateTime(cell.assignmentEndDateTime),
-          assignmentDateTime: moment(cell.assignmentDateTime).format('YYYY-MM-DDTHH:mm:ss'),
-          assignmentEndDateTime: cell.assignmentEndDateTime
-            ? moment(cell.assignmentEndDateTime).format('YYYY-MM-DDTHH:mm:ss')
-            : today,
-          livingUnitId: cell.livingUnitId,
-          agencyId: cell.agencyId,
-        }
-      })
+    const cellData = await Promise.all(
+      cells
+        .sort((left, right) => sortByDateTime(right.assignmentDateTime, left.assignmentDateTime))
+        .filter(cell => cell.assignmentEndDateTime)
+        .map(async cell => {
+          const staffDetails = await prisonApi.getStaffDetails(res.locals, cell.movementMadeBy)
+          const agency = agencyData.find(agencyDetails => cell.agencyId === agencyDetails.agencyId)
+
+          return {
+            establishment: agency.description,
+            location: extractLocation(cell.description, cell.agencyId),
+            movedIn: cell.assignmentDateTime && formatTimestampToDateTime(cell.assignmentDateTime),
+            movedOut: cell.assignmentEndDateTime && formatTimestampToDateTime(cell.assignmentEndDateTime),
+            assignmentDateTime: moment(cell.assignmentDateTime).format('YYYY-MM-DDTHH:mm:ss'),
+            assignmentEndDateTime: cell.assignmentEndDateTime
+              ? moment(cell.assignmentEndDateTime).format('YYYY-MM-DDTHH:mm:ss')
+              : today,
+            livingUnitId: cell.livingUnitId,
+            agencyId: cell.agencyId,
+            movedBy: formatName(staffDetails.firstName, staffDetails.lastName),
+          }
+        })
+    )
 
     return res.render('prisonerProfile/prisonerCellHistory.njk', {
-      cellData,
+      cellHistoryGroupedByAgency: hasLength(cellData)
+        ? Object.entries(groupBy(cellData, 'establishment')).map(([key, value]) => {
+            const fromDateString = value[0].movedIn.split(' -')[0]
```

Comment:
This is potentially a bit fragile if the date format changes, might be worth using moment to format the date time to a date.

Reply:
Agreed.